_posts/2022-05-30-fine-grained-access-control-for-multitenancy.markdown

```diff
@@ -0,0 +1,160 @@
+
+The table provided below is a representation of how the data stored in the purchase history will be partitioned. The parameter `tenantId` is used as the `partition key` while the `itemId` parameter is used as the `sort key`.
+
+|     tenantID     | itemID      |  price   | qty | description |
+| :--------------: | :---------- | :------: | :-: | :---------: |
+| `Customer1-xcv9` | `timestamp` | `number` | `1` |  `string`   |
+| `Customer2-dgf1` | `timestamp` | `number` | `2` |  `string`   |
+| `Customer1-xcv9` | `timestamp` | `number` | `2` |  `string`   |
+| `Customer2-dgf1` | `timestamp` | `number` | `2` |  `string`   |
+
+### Fine Grained Access Control
+
+When the Lambda Authoriser is invoked it is expected to return a JSON object that not only includes a context object that contains the `tenantID` parameter but also a resource policy as detailed [here](#lambda-Authoriser-output-sample). The policy should allow/deny access to the API and dynamoDB table depending on the outcome of the Lambda Authoriser's token claims verification.
+
+All Allow-Policies generated will allow the authenticated user to only access data that has a partition key which matches their tenant id. This is achieved by making use of DynamoDB's `dynamodb:LeadingKeys` condition key for fine grained access control.
+In essence, the `dynamodb:LeadingKeys` condition key is a mechanism for row level access control within DynamodDB tables. Employing the use of this mechanism guarantees a reliable and secure multitenant environment as well as the added benefit of a simplified data persistence layer.
+
+![image](/img/blog/multitenancy/dynamdbFineGrainedAccess.png)
+
+#### Lambda Authoriser Resource Policies
+
+###### Token Verification Failed
+
+```
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Deny",
+            "Action": "*",
+            "Resource": "*"
+        },
+    ]
+}
+```
+
+###### Token Verification Passed
+
+```
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": "execute-api:Invoke",
+            "Resource": executeApiArn,
+        },
+        {
+
+          Effect: 'Allow',
+          Action: ['dynamodb:UpdateItem', 'dynamodb:PutItem', 'dynamodb:DeleteItem', 'dynamodb:Query'],
+          Resource: this.cartTable,
+          Condition: {
+            'ForAllValues:StringLike': {
+              'dynamodb:LeadingKeys': tenantId,
+            },
+          },
+        },
+    ]
+}
+```
+
+#### Wrapping Up.
+
+Congratulations, you have reached the end of the tutorial. If you have any questions or if you think we can help speed up your development journey, please don't hesitate to get in touch with us here at [Mechanical Rock](<(https://www.mechanicalrock.io/lets-get-started/)>).
```

Comment:
Now they have reach the end of your post/tutorial, what should they be able to do? How can they apply it? What else?